modules/explorer/update.go

```diff
@@ -14,36 +11,298 @@ func (e *Explorer) ProcessConsensusChange(cc modules.ConsensusChange) {
+				delete(e.unlockHashes[sfi.UnlockConditions.UnlockHash()], txid)
+				delete(e.unlockHashes[sfi.ClaimUnlockHash], txid)
+				e.siafundInputCount--
+			}
+			for k, sfo := range txn.SiafundOutputs {
+				sfoid := txn.SiafundOutputID(uint64(k))
+				delete(e.siafundOutputIDs[sfoid], txid)
+				delete(e.unlockHashes[sfo.UnlockHash], txid)
+				e.siafundOutputCount--
+			}
+			for _ = range txn.MinerFees {
+				e.minerFeeCount--
+			}
+			for _ = range txn.ArbitraryData {
+				e.arbitraryDataCount--
+			}
+			for _ = range txn.TransactionSignatures {
+				e.transactionSignatureCount--
+			}
 		}
 	}
 
-	// Add the stats for the applied blocks.
+	// Update cumulative stats for applied blocks.
 	for _, block := range cc.AppliedBlocks {
+		// Add the block to the list of active blocks.
+		bid := block.ID()
+		tbid := types.TransactionID(bid)
+		e.blockchainHeight++
+		e.blockHashes[bid] = e.blockchainHeight
+		e.transactionHashes[tbid] = e.blockchainHeight // Miner payouts are a transaciton.
+
+		// Catalog the new miner payouts.
+		for j, payout := range block.MinerPayouts {
+			scoid := block.MinerPayoutID(uint64(j))
+			_, exists := e.siacoinOutputIDs[scoid]
+			if !exists {
+				e.siacoinOutputIDs[scoid] = make(map[types.TransactionID]struct{})
+			}
+			e.siacoinOutputIDs[scoid][tbid] = struct{}{}
+			_, exists = e.unlockHashes[payout.UnlockHash]
+			if !exists {
+				e.unlockHashes[payout.UnlockHash] = make(map[types.TransactionID]struct{})
+			}
+			e.unlockHashes[payout.UnlockHash][tbid] = struct{}{}
+		}
+
+		// Update cumulative stats for applied transactions.
 		for _, txn := range block.Transactions {
-			// Revert all of the file contracts.
-			for _, fc := range txn.FileContracts {
-				e.totalContractCount += 1
+			// Add the transaction to the list of active transactions.
+			txid := txn.ID()
+			e.transactionCount++
+			e.transactionHashes[txid] = e.blockchainHeight
+
+			for _, sci := range txn.SiacoinInputs {
+				_, exists := e.siacoinOutputIDs[sci.ParentID]
+				if !exists {
+					panic("siacoin input without siacoin output")
+				}
+				e.siacoinOutputIDs[sci.ParentID][txid] = struct{}{}
+				_, exists = e.unlockHashes[sci.UnlockConditions.UnlockHash()]
+				if !exists {
+					panic("unlock conditions without a parent unlock hash")
+				}
+				e.unlockHashes[sci.UnlockConditions.UnlockHash()][txid] = struct{}{}
+				e.siacoinInputCount++
+			}
+			for j, sco := range txn.SiacoinOutputs {
+				scoid := txn.SiacoinOutputID(uint64(j))
+				_, exists := e.siacoinOutputIDs[scoid]
+				if !exists {
+					e.siacoinOutputIDs[scoid] = make(map[types.TransactionID]struct{})
+				}
+				e.siacoinOutputIDs[scoid][txid] = struct{}{}
+				_, exists = e.unlockHashes[sco.UnlockHash]
+				if !exists {
+					e.unlockHashes[sco.UnlockHash] = make(map[types.TransactionID]struct{})
+				}
+				e.unlockHashes[sco.UnlockHash][txn.ID()] = struct{}{}
+				e.siacoinOutputCount++
+			}
+			for k, fc := range txn.FileContracts {
```

Comment:
4, I get all the way to 'l' in some of the loops.

I probably should break this into more functions.